### PR TITLE
Gracefully handle non-ASCII chars in dataset names 🌶️

### PR DESCRIPTION
**Pull request by deepyaman**
## Description

* Have simpler, yet more robust, handling around the "friendly" conversion of data set names
* It also helps avoid clashes like between datasets "perchè" and "perché" or even "jalapeño" and "jalape@o"
* On the last two points, if you do want special handling of accents, it should be more explicit/not here

## Development notes
Dropped a Python 3.5 Windows relic allowing only ASCII files

## Checklist

- [x] Read the [contributing](https://github.com/quantumblacklabs/kedro/blob/master/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](https://github.com/quantumblacklabs/kedro/blob/master/RELEASE.md) file
- [x] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: tests/io/test_data_catalog.py [hunk 1]
```diff
@@ -628,12 +628,15 @@ def test_load_version_on_unversioned_dataset(
     def test_replacing_non_alphanumeric_characters(self):
```
Comment (921kiyo) [inline tests/io/test_data_catalog.py:628]: And this test function name and docstring as well.

File: kedro/io/data_catalog.py [hunk 2]
```diff
@@ -116,10 +116,7 @@ class _FrozenDatasets:
     def __init__(self, datasets):
         # Non-alphanumeric characters (except underscore) in dataset name
```
Comment (921kiyo) [inline kedro/io/data_catalog.py:117]: It says replace non-alphanumeric characters, but `ñ` in `jalapeño` is a non-alphanumeric character. So this needs to be updated.

Comment (deepyaman) [reply kedro/io/data_catalog.py:117]: @921kiyo I can (and considered) change it to read "non-word characters", but I'd argue ñ is part of the Spanish alphabet. :)

Comment (921kiyo) [reply kedro/io/data_catalog.py]: I think it will work for Chinese or Japanese characters (non-alphabet)? :)

**Comment by deepyaman** [comment]
One more question: Would it make sense to move this into a function?

The reason I came across this was because `0.16.2` broke my code, so I'm handling it like so: https://github.com/deepyaman/kedro-accelerator/commit/86897b039368df8a134db233267141e26aedad85

Having it as a function would mean less likely changes for my (and similar) code, although I'm not sure this should change that frequently anyway.

**Comment by 921kiyo** [comment]
Yeah, you can wrap `datasets = {re.sub(r"\W+", "__", key): value for key, value in datasets.items()}` as a private function, or leave it as it. Since it is one line, I'm happy with either.

**Comment by deepyaman** [comment]
> Yeah, you can wrap `datasets = {re.sub(r"\W+", "__", key): value for key, value in datasets.items()}` as a private function, or leave it as it. Since it is one line, I'm happy with either.

Sorry, didn't do it last night as promised. I'll do it shortly

**Pull request merged**